### PR TITLE
Meson build system for ratbagd

### PR DESCRIPTION
**Pull request by whot**
This is the switch to use meson instead of autotools. Note that we need tmeson 0.40 for the python bits and that was only released a few days ago.

This contains the two patches also in PR #15

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,164 @@
+project('ratbagd', 'c',
+	version : '0.2',
+	license : 'MIT/Expat',
+	meson_version : '>= 0.40.0')
```
Comment (bentiss) [inline meson.build]: don't we need to add `default_options : [ 'c_std=gnu99' ]` to keep it in sync with libratbag?

File: meson.build [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        '-fdiagnostics-show-option',
+        '-fno-strict-aliasing',
+        '-fvisibility=hidden',
+        '-ffunction-sections',
+        '-fdata-sections',
+        '-fstack-protector',
+        '-fstack-protector-strong',
+        '-fPIE',
+        '--param=ssp-buffer-size=4'
+]
```
Comment (bentiss) [inline meson.build]: Why suddenly have this extensive list of compiler flags when libratbag doesn't?

Comment (bentiss) [reply meson.build]: Well, I guess they were in the autotools already...

File: .gitignore [hunk 3]
```diff
@@ -1,36 +1,13 @@
...
-*.m4
-!m4/attributes.m4
-ratbagctl
-./ratbagd
-stamp-*
-ratbagd
-ratbagd.service
-ratbagctl
-org.freedesktop.ratbag1.service
+build/
```
Comment (bentiss) [inline .gitignore]: builddir?

**Review by bentiss — COMMENTED** [review]
Few minor issues, nothing significant (except the .gitignore maybe)

File: .gitignore (hunk 3 above)
Comment (dvdhrm) [reply .gitignore]: I use empty gitignores these days with meson. The builddir is picked by everyone individually, so they can either accept it showing up in `git status`, or they can just put it into `~/.gitconfig` (or whatever it is called).

But this just as a side note... I don't see any problem with keeping the entries if needed.

File: README.md [hunk 4]
```diff
@@ -113,31 +113,24 @@ activation is possible.
 Compiling ratbagd
 -----------------
 
-ratbagd needs systemd >= 227. If you are running Fedora 23, you might find the
-following instruction valuable:
+ratbagd uses the meson build system, see http://mesonbuild.com which in
```
Comment (dvdhrm) [inline README.md]: The comma looks misplaced. I'd use:

"ratbagd uses the meson build system (see http://...), which in..."

...but maybe that's just my German grammar showing through.

File: README.md [hunk 5]
```diff
@@ -113,31 +113,24 @@ activation is possible.
...
+turn uses ninja to invoke the compiler (`ninja` may be `ninja-build` on your
+distribution). From a fresh git checkout, run the
+following commands to init the repository:
```
Comment (dvdhrm) [inline README.md]: Why the line-break after `run the`, rather than aligning everything the same?

File: meson.build [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+cc = meson.get_compiler('c')
+cflags = [
+        '-pipe',
+        '-Wall',
+        '-Wextra',
+        '-Wno-inline',
+        '-Wundef',
+        '-Wformat=2',
+	'-Wformat-security',
+	'-Wformat-nonliteral',
```
Comment (dvdhrm) [inline meson.build]: Indentation?

File: meson.build (hunk 2 above)
Comment (dvdhrm) [reply meson.build]: Legacy, that originated in systemd code... The only ones I use these days are `-fno-common` and `-fvisibility=hidden`. I think everything else can be safely dropped. Especially the warnings belong into the environment of the developer, not the upstream build instructions.

File: meson.build [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        '-fstack-protector-strong',
+        '-fPIE',
+        '--param=ssp-buffer-size=4'
+]
+add_project_arguments(cflags, language: 'c')
+
+# Initialize config.h, to be added to in the various options below, config.h
+# is generated at the end of this file
+config_h = configuration_data()
+config_h.set('_GNU_SOURCE', '1')
```
Comment (dvdhrm) [inline meson.build]: I prefer:
    `add_project_arguments('-D_GNU_SOURCE', language: 'c')`

However, no strong feelings here. I just think it is weird to make this a `configuration` setting, when it is actually something that selects your build environment. Furthermore, when in `config.h`, you must make sure to include it before including any other system header file (which then most projects work around via the `-include config.h` gcc cmdline argument).

Up to you..

**Review by dvdhrm — APPROVED** [review]
Only minor nits. Looks nice, thanks a lot!

File: meson.build (hunk 7 above)
Comment (whot) [reply meson.build]: I think it's just the direct translation from automake where it was placed into config.h as well, but I'm happy to change this.

**Comment by whot** [comment]
weird, I can't reply to the "legacy that originated" commend, so I'll just comment here: @dvdhrm, I *mostly* agree about the warnings being in the developer environment but how do you deal with developers that don't have these set and thus introduce warnings? It requires an extra step from the maintainers to make sure they aren't triggered, either in the CI system or during review (i.e. build on local machine)

File: meson.build (hunk 2 above)
Comment (whot) [reply meson.build]: planning to leave this in for the current conversion, we can trim it later

**Comment by whot** [comment]
force-pushed, all comments addresses as expected, except for the cflags that I left in place

**Review by bentiss — APPROVED** [review]
Looks good to me

**Comment by dvdhrm** [comment]
> I mostly agree about the warnings being in the developer environment but how do you deal with developers that don't have these set and thus introduce warnings? It requires an extra step from the maintainers to make sure they aren't triggered, either in the CI system or during review (i.e. build on local machine)

The same is true if they use ancient compilers, or if their local environment contains `-Wno-foobar`, or if they simply ignore warnings.

I have no solution for this. I just decided for myself that it is not worth my time maintaining all those compiler warnings upstream. But I am not going fight for their removal ;)

**Pull request merged**